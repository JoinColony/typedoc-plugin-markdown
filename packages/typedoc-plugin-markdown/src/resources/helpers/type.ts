import * as Handlebars from 'handlebars/runtime';
import {
  ArrayType,
  ConditionalType,
  DeclarationReflection,
  IndexedAccessType,
  InferredType,
  IntersectionType,
  IntrinsicType,
  LiteralType,
  PredicateType,
  QueryType,
  ReferenceType,
  ReflectionType,
  SignatureReflection,
  TupleType,
  TypeOperatorType,
  UnionType,
  UnknownType,
} from 'typedoc';

type Collapse = 'object' | 'function' | 'all' | 'none';

export const type = (
  context:
    | ArrayType
    | IntersectionType
    | IntrinsicType
    | ReferenceType
    | TupleType
    | UnionType
    | TypeOperatorType
    | QueryType
    | PredicateType
    | ReferenceType
    | ConditionalType
    | IndexedAccessType
    | UnknownType
    | InferredType,
  collapse: Collapse = 'none',
  emphasis = true,
) => {
  if (context instanceof ReferenceType) {
    return getReferenceType(context, emphasis);
  }

  if (context instanceof ArrayType && context.elementType) {
    return getArrayType(context, emphasis);
  }

  if (context instanceof UnionType && context.types) {
    return getUnionType(context, emphasis);
  }

  if (context instanceof IntersectionType && context.types) {
    return getIntersectionType(context);
  }

  if (context instanceof TupleType && context.elements) {
    return getTupleType(context);
  }

  if (context instanceof IntrinsicType && context.name) {
    return getIntrinsicType(context, emphasis);
  }

  if (context instanceof ReflectionType) {
    return getReflectionType(context, collapse);
  }

  if (context instanceof DeclarationReflection) {
    return getReflectionType(context, collapse);
  }

  if (context instanceof TypeOperatorType) {
    return getTypeOperatorType(context);
  }

  if (context instanceof QueryType) {
    return getQueryType(context);
  }

  if (context instanceof ConditionalType) {
    return getConditionalType(context);
  }

  if (context instanceof IndexedAccessType) {
    return getIndexAccessType(context);
  }

  if (context instanceof UnknownType) {
    return getUnknownType(context);
  }

  if (context instanceof InferredType) {
    return getInferredType(context);
  }

  if (context instanceof LiteralType) {
    return getLiteralType(context);
  }

  return context ? Handlebars.helpers.escapeChars(context.toString()) : '';
};

//}

function getLiteralType(model: LiteralType) {
  if (typeof model.value === 'bigint') {
    return `\`${model.value}n\``;
  }
  return `\`\`${JSON.stringify(model.value)}\`\``;
}

export function getReflectionType(
  context: DeclarationReflection | ReflectionType,
  collapse: Collapse,
) {
  const root =
    context instanceof ReflectionType ? context.declaration : context;
  if (root.signatures) {
    return collapse === 'function' || collapse === 'all'
      ? `\`fn\``
      : getFunctionType(root.signatures);
  }
  return collapse === 'object' || collapse === 'all'
    ? `\`Object\``
    : getDeclarationType(root);
}

function getDeclarationType(model: DeclarationReflection) {
  if (model.indexSignature || model.children) {
    let indexSignature = '';
    const declarationIndexSignature = model.indexSignature;
    if (declarationIndexSignature) {
      const key = declarationIndexSignature.parameters
        ? declarationIndexSignature.parameters.map(
            (param) => `\`[${param.name}: ${param.type}]\``,
          )
        : '';
      const obj = Handlebars.helpers.type(declarationIndexSignature.type);
      indexSignature = `${key}: ${obj}; `;
    }
    const types =
      model.children &&
      model.children.map((obj) => {
        return `\`${obj.name}${
          obj.flags.isOptional ? '?' : ''
        }\`: ${Handlebars.helpers.type(
          obj.signatures || obj.children ? obj : obj.type,
        )} ${
          obj.defaultValue && obj.defaultValue !== '...'
            ? `= ${Handlebars.helpers.escapeChars(obj.defaultValue)}`
            : ''
        }`;
      });
    return `{ ${indexSignature ? indexSignature : ''}${
      types ? types.join('; ') : ''
    } }${
      model.defaultValue && model.defaultValue !== '...'
        ? `= ${Handlebars.helpers.escapeChars(model.defaultValue)}`
        : ''
    }`;
  }
  return '{}';
}

export function getFunctionType(modelSignatures: SignatureReflection[]) {
  const functions = modelSignatures.map((fn) => {
    const typeParams = fn.typeParameters
      ? `<${fn.typeParameters
          .map((typeParameter) => typeParameter.name)
          .join(', ')}\\>`
      : [];
    const params = fn.parameters
      ? fn.parameters.map((param) => {
          return `${param.flags.isRest ? '...' : ''}\`${param.name}${
            param.flags.isOptional ? '?' : ''
          }\`: ${Handlebars.helpers.type(param.type ? param.type : param)}`;
        })
      : [];
    const returns = Handlebars.helpers.type(fn.type);
    return typeParams + `(${params.join(', ')}) => ${returns}`;
  });
  return functions.join('');
}

function getReferenceType(model: ReferenceType, emphasis) {
  //const externalUrl = context.attemptExternalResolution(model);
  const externalUrl = false;
  if (model.reflection || (model.name && model.typeArguments)) {
    const reflection: string[] = [];

    if (model.reflection?.url) {
      reflection.push(
        `[${`\`${model.reflection.name}\``}](${Handlebars.helpers.relativeURL(
          model.reflection.url,
        )})`,
      );
    } else {
      reflection.push(
        externalUrl
          ? `[${`\`${model.name}\``}]( ${externalUrl} )`
          : `\`${model.name}\``,
      );
    }
    if (model.typeArguments && model.typeArguments.length > 0) {
      reflection.push(
        `<${model.typeArguments
          .map((typeArgument) => Handlebars.helpers.type(typeArgument))
          .join(', ')}\\>`,
      );
    }
    return reflection.join('');
  }
  return emphasis
    ? externalUrl
      ? `[${`\`${model.name}\``}]( ${externalUrl} )`
      : `\`${model.name}\``
    : Handlebars.helpers.escapeChars(model.name);
}

function getArrayType(model: ArrayType, emphasis: boolean) {
  const arrayType = Handlebars.helpers.type(
    model.elementType,
    'none',
    emphasis,
  );
  return model.elementType.type === 'union'
    ? `(${arrayType})[]`
    : `${arrayType}[]`;
}

function getUnionType(model: UnionType, emphasis: boolean) {
  return model.types
    .map((unionType) => Handlebars.helpers.type(unionType, 'none', emphasis))
    .join(` \\| `);
}

function getIntersectionType(model: IntersectionType) {
  return model.types
    .map((intersectionType) => Handlebars.helpers.type(intersectionType))
    .join(' & ');
}

function getTupleType(model: TupleType) {
  return `[${model.elements
    .map((element) => Handlebars.helpers.type(element))
    .join(', ')}]`;
}

function getIntrinsicType(model: IntrinsicType, emphasis: boolean) {
  return emphasis
    ? `\`${model.name}\``
    : Handlebars.helpers.escapeChars(model.name);
}

function getTypeOperatorType(model: TypeOperatorType) {
  return `${model.operator} ${Handlebars.helpers.type(model.target)}`;
}

function getQueryType(model: QueryType) {
  return `typeof ${Handlebars.helpers.type(model.queryType)}`;
}

function getInferredType(model: InferredType) {
  return `infer ${Handlebars.helpers.escapeChars(model.name)}`;
}

function getUnknownType(model: UnknownType) {
  return Handlebars.helpers.escapeChars(model.name);
}

function getConditionalType(model: ConditionalType) {
  const md: string[] = [];
  if (model.checkType) {
    md.push(Handlebars.helpers.type(model.checkType));
  }
  md.push('extends');
  if (model.extendsType) {
    md.push(Handlebars.helpers.type(model.extendsType));
  }
  md.push('?');
  if (model.trueType) {
    md.push(Handlebars.helpers.type(model.trueType));
  }
  md.push(':');
  if (model.falseType) {
    md.push(Handlebars.helpers.type(model.falseType));
  }
  return md.join(' ');
}

function getIndexAccessType(model: IndexedAccessType) {
  const md: string[] = [];
  if (model.objectType) {
    md.push(Handlebars.helpers.type(model.objectType));
  }
  if (model.indexType) {
    md.push(`[${Handlebars.helpers.type(model.indexType)}]`);
  }
  return md.join('');
}
