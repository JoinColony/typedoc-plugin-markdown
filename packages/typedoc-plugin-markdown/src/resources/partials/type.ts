import {
  ArrayType,
  ConditionalType,
  DeclarationReflection,
  IndexedAccessType,
  InferredType,
  IntersectionType,
  IntrinsicType,
  LiteralType,
  PredicateType,
  QueryType,
  ReferenceType,
  ReflectionType,
  SignatureReflection,
  TupleType,
  TypeOperatorType,
  UnionType,
  UnknownType,
} from 'typedoc';
import { MarkdownThemeContext } from '../../theme-context';
import { escapeChars } from '../../utils/format';

type Collapse = 'object' | 'function' | 'all' | 'none';

export const typePartial = (
  context: MarkdownThemeContext,
  props:
    | ArrayType
    | IntersectionType
    | IntrinsicType
    | ReferenceType
    | TupleType
    | UnionType
    | TypeOperatorType
    | QueryType
    | PredicateType
    | ReferenceType
    | ConditionalType
    | IndexedAccessType
    | UnknownType
    | InferredType,

  collapse: Collapse = 'none',
  emphasis = true,
) => {
  function getLiteralType(model: LiteralType) {
    if (typeof model.value === 'bigint') {
      return `\`${model.value}n\``;
    }
    return `\`\`${JSON.stringify(model.value)}\`\``;
  }

  function getReflectionType(
    model: DeclarationReflection | ReflectionType,
    collapse: Collapse,
  ) {
    const root = model instanceof ReflectionType ? model.declaration : model;
    if (root.signatures) {
      return getFunctionType(root.signatures, collapse);
    }
    return collapse === 'object' || collapse === 'all'
      ? `\`Object\``
      : getDeclarationType(root);
  }

  function getDeclarationType(model: DeclarationReflection) {
    if (model.indexSignature || model.children) {
      let indexSignature = '';
      const declarationIndexSignature = model.indexSignature;
      if (declarationIndexSignature) {
        const key = declarationIndexSignature.parameters
          ? declarationIndexSignature.parameters.map(
              (param) => `[${param.name}: ${param.type}]`,
            )
          : '';
        const obj = context.typePartial(declarationIndexSignature.type);
        indexSignature = `${key}: ${obj}; `;
      }
      const types =
        model.children &&
        model.children.map((obj) => {
          return `\`${obj.name}${
            obj.flags.isOptional ? '?' : ''
          }\`: ${context.typePartial(
            obj.signatures || obj.children ? obj : obj.type,
          )} ${
            obj.defaultValue && obj.defaultValue !== '...'
              ? `= ${escapeChars(obj.defaultValue)}`
              : ''
          }`;
        });
      return `{ ${indexSignature ? indexSignature : ''}${
        types ? types.join('; ') : ''
      } }${
        model.defaultValue && model.defaultValue !== '...'
          ? `= ${escapeChars(model.defaultValue)}`
          : ''
      }`;
    }
    return '{}';
  }

  function getFunctionType(
    modelSignatures: SignatureReflection[],
    collapse: Collapse,
  ) {
    if (collapse === 'function' || collapse === 'all') {
      return `\`fn\``;
    }
    const functions = modelSignatures.map((fn) => {
      const typeParams = fn.typeParameters
        ? `<${fn.typeParameters
            .map((typeParameter) => typeParameter.name)
            .join(', ')}\\>`
        : [];
      const params = fn.parameters
        ? fn.parameters.map((param) => {
            return `${param.flags.isRest ? '...' : ''}\`${param.name}${
              param.flags.isOptional ? '?' : ''
            }\`: ${context.typePartial(param.type ? param.type : param)}`;
          })
        : [];
      const returns = context.typePartial(fn.type);
      return typeParams + `(${params.join(', ')}) => ${returns}`;
    });
    return functions.join('');
  }

  function getReferenceType(model: ReferenceType, emphasis: boolean) {
    const externalUrl = context.attemptExternalResolution(model);
    if (model.reflection || (model.name && model.typeArguments)) {
      const reflection: string[] = [];

      if (model.reflection?.url) {
        reflection.push(
          `[${`\`${model.reflection.name}\``}](${context.urlTo(
            model.reflection,
          )})`,
        );
      } else {
        reflection.push(
          externalUrl
            ? `[${`\`${model.name}\``}]( ${externalUrl} )`
            : `\`${model.name}\``,
        );
      }
      if (model.typeArguments && model.typeArguments.length > 0) {
        reflection.push(
          `<${model.typeArguments
            .map((typeArgument) => context.typePartial(typeArgument))
            .join(', ')}\\>`,
        );
      }
      return reflection.join('');
    }
    return emphasis
      ? externalUrl
        ? `[${`\`${model.name}\``}]( ${externalUrl} )`
        : `\`${model.name}\``
      : escapeChars(model.name);
  }

  function getArrayType(model: ArrayType, emphasis: boolean) {
    const arrayType = context.typePartial(model.elementType, 'none', emphasis);
    return model.elementType.type === 'union'
      ? `(${arrayType})[]`
      : `${arrayType}[]`;
  }

  function getUnionType(model: UnionType, emphasis: boolean) {
    return model.types
      .map((unionType) => context.typePartial(unionType, 'none', emphasis))
      .join(` \\| `);
  }

  function getIntersectionType(model: IntersectionType) {
    return model.types
      .map((intersectionType) => context.typePartial(intersectionType))
      .join(' & ');
  }

  function getTupleType(model: TupleType) {
    return `[${model.elements
      .map((element) => context.typePartial(element))
      .join(', ')}]`;
  }

  function getIntrinsicType(model: IntrinsicType, emphasis: boolean) {
    return emphasis ? `\`${model.name}\`` : escapeChars(model.name);
  }

  function getTypeOperatorType(model: TypeOperatorType) {
    return `${model.operator} ${context.typePartial(model.target)}`;
  }

  function getQueryType(model: QueryType) {
    return `typeof ${context.typePartial(model.queryType)}`;
  }

  function getInferredType(model: InferredType) {
    return `infer ${escapeChars(model.name)}`;
  }

  function getUnknownType(model: UnknownType) {
    return escapeChars(model.name);
  }

  function getConditionalType(model: ConditionalType) {
    const md: string[] = [];
    if (model.checkType) {
      md.push(context.typePartial(model.checkType));
    }
    md.push('extends');
    if (model.extendsType) {
      md.push(context.typePartial(model.extendsType));
    }
    md.push('?');
    if (model.trueType) {
      md.push(context.typePartial(model.trueType));
    }
    md.push(':');
    if (model.falseType) {
      md.push(context.typePartial(model.falseType));
    }
    return md.join(' ');
  }

  function getIndexAccessType(model: IndexedAccessType) {
    const md: string[] = [];
    if (model.objectType) {
      md.push(context.typePartial(model.objectType));
    }
    if (model.indexType) {
      md.push(`[${context.typePartial(model.indexType)}]`);
    }
    return md.join('');
  }

  if (props instanceof ReferenceType) {
    return getReferenceType(props, emphasis);
  }

  if (props instanceof ArrayType && props.elementType) {
    return getArrayType(props, emphasis);
  }

  if (props instanceof UnionType && props.types) {
    return getUnionType(props, emphasis);
  }

  if (props instanceof IntersectionType && props.types) {
    return getIntersectionType(props);
  }

  if (props instanceof TupleType && props.elements) {
    return getTupleType(props);
  }

  if (props instanceof IntrinsicType && props.name) {
    return getIntrinsicType(props, emphasis);
  }

  if (props instanceof ReflectionType) {
    return getReflectionType(props, collapse);
  }

  if (props instanceof DeclarationReflection) {
    return getReflectionType(props, collapse);
  }

  if (props instanceof Array && props[0] instanceof SignatureReflection) {
    return getFunctionType(props, collapse);
  }

  if (props instanceof TypeOperatorType) {
    return getTypeOperatorType(props);
  }

  if (props instanceof QueryType) {
    return getQueryType(props);
  }

  if (props instanceof ConditionalType) {
    return getConditionalType(props);
  }

  if (props instanceof IndexedAccessType) {
    return getIndexAccessType(props);
  }

  if (props instanceof UnknownType) {
    return getUnknownType(props);
  }

  if (props instanceof InferredType) {
    return getInferredType(props);
  }

  if (props instanceof LiteralType) {
    return getLiteralType(props);
  }

  return props ? escapeChars(props.toString()) : '';
};
