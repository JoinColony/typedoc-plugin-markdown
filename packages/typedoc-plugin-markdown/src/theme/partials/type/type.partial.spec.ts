import { ProjectReflection } from 'typedoc';
import { typePartial } from './type.partial';

describe(`typePartial`, () => {
  let project: ProjectReflection;

  beforeAll(async () => {
    project = global.getProject('types.ts');
  });

  test(`should compile 'array' type'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('arrayType') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile 'stringLiteral' type'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('stringLiteralType') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile 'union' of string literals types'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('unionType') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile 'union' of literal declarations`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (
          project.findReflectionByName(
            'unionTypeWithSymbolsDeclarations',
          ) as any
        ).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile intrinsic type'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('stringType') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile collapsed 'literal' type'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('literalType') as any).type,
        'all',
      ),
    ).toMatchSnapshot();
  });

  test(`should compile expanded 'literal' type'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('literalType') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile collapsed 'objectLiteralType' type'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        project.findReflectionByName('objectLiteralType') as any,
        'object',
      ),
    ).toMatchSnapshot();
  });

  test(`should compile expanded 'objectLiteralType' type'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        project.findReflectionByName('objectLiteralType') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile 'tuple' type'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('tupleType') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile 'intersection' type'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('intersectionType') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile collapsed 'function' type '`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('functionReflectionType') as any).type,
        'function',
      ),
    ).toMatchSnapshot();
  });

  test(`should compile expanded 'function' type '`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('functionReflectionType') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile 'typeOperator' type '`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('typeOperatorType') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile unionType with object literal type '`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('objectLiteralUnionType') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile conditional type '`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('ConditionalType') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should resolve external refs'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (project.findReflectionByName('externalReference') as any).type,
      ),
    ).toMatchSnapshot();
  });

  test(`should resolve external refs with type params'`, () => {
    expect(
      typePartial(
        global.getMockContext(),
        (
          project.findReflectionByName(
            'externalReferenceInsideTypeParams',
          ) as any
        ).type,
      ),
    ).toMatchSnapshot();
  });
});
