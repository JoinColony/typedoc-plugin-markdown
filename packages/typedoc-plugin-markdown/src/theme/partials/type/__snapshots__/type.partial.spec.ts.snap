// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typePartial should compile 'array' type' 1`] = `"\`string\`[]"`;

exports[`typePartial should compile 'intersection' type' 1`] = `"[\`IntersectionClassA\`]( IntersectionClassA ) & [\`IntersectionClassB\`]( IntersectionClassB )"`;

exports[`typePartial should compile 'stringLiteral' type' 1`] = `"\`\`\\"blue\\"\`\`"`;

exports[`typePartial should compile 'tuple' type' 1`] = `"[\`string\`, \`number\`]"`;

exports[`typePartial should compile 'typeOperator' type ' 1`] = `"unique \`symbol\`"`;

exports[`typePartial should compile 'union' of literal declarations 1`] = `"{ \`bar\`: \`\`\\" \\"\`\` \\\\| \`\`\\"string\\"\`\` \\\\| \`\`\\"strong|with|pipes\\"\`\` \\\\| \`\`\\"type\`with\`backticks\\"\`\` \\\\| \`\`\\"*\\"\`\`  }"`;

exports[`typePartial should compile 'union' of string literals types' 1`] = `"\`\`\\"ease-in\\"\`\` \\\\| \`\`\\"ease-out\\"\`\`"`;

exports[`typePartial should compile collapsed 'function' type ' 1`] = `""`;

exports[`typePartial should compile collapsed 'literal' type' 1`] = `"\`Object\`"`;

exports[`typePartial should compile collapsed 'objectLiteralType' type' 1`] = `"\`Object\`"`;

exports[`typePartial should compile conditional type ' 1`] = `"[\`T\`]( T ) extends \`string\` ? \`\`\\"string\\"\`\` : [\`T\`]( T ) extends \`number\` ? \`\`\\"number\\"\`\` : [\`T\`]( T ) extends \`boolean\` ? \`\`\\"boolean\\"\`\` : [\`T\`]( T ) extends \`undefined\` ? \`\`\\"undefined\\"\`\` : \`\`\\"object\\"\`\`"`;

exports[`typePartial should compile expanded 'function' type ' 1`] = `""`;

exports[`typePartial should compile expanded 'literal' type' 1`] = `"{ \`valueA?\`: \`number\` ; \`valueB?\`: \`boolean\` ; \`valueX\`: { \`valueA\`: \`number\`[] ; \`valueY\`: (\`z\`: \`string\`) => { \`a\`: \`string\` ; \`b\`: \`string\`  } ; \`valueZ\`: \`string\`  } ; \`valueY\`: () => \`string\` ; \`valueZ\`: \`string\`  }"`;

exports[`typePartial should compile expanded 'objectLiteralType' type' 1`] = `"{}"`;

exports[`typePartial should compile intrinsic type' 1`] = `"\`string\`"`;

exports[`typePartial should compile unionType with object literal type ' 1`] = `"\`string\` \\\\| { \`z\`: \`string\`  }"`;

exports[`typePartial should resolve external refs with type params' 1`] = `"[\`ClassWithTypeParams\`]( ClassWithTypeParams )<[\`HTMLElement\`]( HTMLElement ), [\`Error\`]( Error )\\\\>"`;

exports[`typePartial should resolve external refs' 1`] = `"[\`HTMLElement\`]( HTMLElement )"`;
